Firestore/core/src/firebase/firestore/remote/grpc_stream.mm

```diff
@@ -0,0 +1,368 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "Firestore/core/src/firebase/firestore/remote/grpc_stream.h"
+
+#include <grpcpp/create_channel.h>
+#include <utility>
+
+#include "Firestore/core/src/firebase/firestore/util/executor_libdispatch.h"
+#include "Firestore/core/src/firebase/firestore/util/firebase_assert.h"
+#include "Firestore/Source/Remote/FSTDatastore.h"
+#include "Firestore/Source/Remote/FSTStream.h"
+#include "absl/memory/memory.h"
+
+#include <fstream>
+#include <sstream>
+
+namespace firebase {
+namespace firestore {
+namespace remote {
+
+using firebase::firestore::model::SnapshotVersion;
+
+namespace {
+
+enum class GrpcOperationTag {
+  Start,
+  Read,
+  Write,
+  Finish,
+};
+
+GrpcOperationTag kStartTag = GrpcOperationTag::Start;
+GrpcOperationTag kReadTag = GrpcOperationTag::Read;
+GrpcOperationTag kWriteTag = GrpcOperationTag::Write;
+GrpcOperationTag kFinishTag = GrpcOperationTag::Finish;
+
+}  // namespace
+
+namespace internal {
+
+std::unique_ptr<grpc::ClientContext> ObjcBridge::CreateContext(
+    const model::DatabaseId& database_id, const absl::string_view token) const {
+  return [FSTDatastore createGrpcClientContextWithDatabaseID:&database_id
+                                                       token:token];
+}
+
+grpc::ByteBuffer ObjcBridge::ToByteBuffer(FSTQueryData* query) const {
+  GCFSListenRequest* request = [GCFSListenRequest message];
+  request.database = [serializer_ encodedDatabaseID];
+  request.addTarget = [serializer_ encodedTarget:query];
+  request.labels = [serializer_ encodedListenRequestLabelsForQueryData:query];
+
+  return ToByteBuffer([request data]);
+}
+
+grpc::ByteBuffer ObjcBridge::ToByteBuffer(FSTTargetID target_id) const {
+  GCFSListenRequest* request = [GCFSListenRequest message];
+  request.database = [serializer_ encodedDatabaseID];
+  request.removeTarget = target_id;
+
+  return ToByteBuffer([request data]);
+}
+
+grpc::ByteBuffer ObjcBridge::ToByteBuffer(NSData* data) const {
+  const grpc::Slice slice{[data bytes], [data length]};
+  return grpc::ByteBuffer{&slice, 1};
+}
+
+FSTWatchChange* ObjcBridge::GetWatchChange(GCFSListenResponse* proto) {
+  return [serializer_ decodedWatchChange:proto];
+}
+
+SnapshotVersion ObjcBridge::GetSnapshotVersion(GCFSListenResponse* proto) {
+  return [serializer_ versionFromListenResponse:proto];
+}
+
+NSData* ObjcBridge::ToNsData(const grpc::ByteBuffer& buffer) const {
+  std::vector<grpc::Slice> slices;
+  buffer.Dump(&slices);  // TODO: check return value
+  if (slices.size() == 1) {
+    return [NSData dataWithBytes:slices.front().begin()
+                          length:slices.front().size()];
+  } else {
+    // FIXME
+    return [NSData dataWithBytes:slices.front().begin()
```

Comment:
Maybe `abort()`? If there's more than one slice here this would silently corrupt.